File: JulyLongChallenge2018/NMNMXtemp.py

#include <stdio.h>
#include <stdlib.h>

mod = int(1e9)+7
T = int(input())

def modInv(a,m=1000000007):
    m0=m
    y=0
    x=1
    if m==1:
        return 0
    while (a>1):
        q=a//m
        t=m
        m=a%m
        a=t
        t=y
        y=x-q*y
        x=t

    if (x<0):
        x+=m0
    return x


def modDivide(a,b,m=1000000007):
    a=a%m
    inv=modInv(b,m)
    return (inv*a)%m




# A Dynamic Programming based Python Program that uses table C[][]
# to calculate the Binomial Coefficient

# Returns value of Binomial Coefficient C(n, k)
def nCr(n, k,p):
    C = [[0 for x in range(k+1)] for x in range(n+1)]

    # Calculate value of Binomial Coefficient in bottom up manner
    for i in range(n+1):
        for j in range(min(i, k)+1):
            # Base Cases
            if j == 0 or j == i:
                C[i][j] = 1

            # Calculate value using previosly stored values
            else:
                C[i][j] = ((C[i-1][j-1])%(p-1) + (C[i-1][j])%(p-1))% (p-1)

    return C[n][k]% (p-1)

def power(x, y, p) :
    res = 1     # Initialize result


    x = x % p

    while (y > 0) :

        if ((y & 1) == 1) :
            res = (res * x) % p


        y = y >> 1      # y = y/2
        x = (x * x) % p

    return res % p

# This function takes last element as pivot, places
# the pivot element at its correct position in sorted
# array, and places all smaller (smaller than pivot)
# to left of pivot and all greater elements to right
# of pivot
def partition(arr,low,high):
    i = ( low-1 )         # index of smaller element
    pivot = arr[high]     # pivot

    for j in range(low , high):

        # If current element is smaller than or
        # equal to pivot
        if   arr[j] <= pivot:

            # increment index of smaller element
            i = i+1
            arr[i],arr[j] = arr[j],arr[i]

    arr[i+1],arr[high] = arr[high],arr[i+1]
    return ( i+1 )

# The main function that implements QuickSort
# arr[] --> Array to be sorted,
# low  --> Starting index,
# high  --> Ending index

# Function to do Quick sort
def quickSort(arr,low,high):
    if low < high:

        # pi is partitioning index, arr[p] is now
        # at right place
        pi = partition(arr,low,high)

        # Separately sort elements before
        # partition and after partition
        quickSort(arr, low, pi-1)
        quickSort(arr, pi+1, high)



for test_case in range(T):
    N,K=map(int,input().split())

    seq = list(map(int,input().split()))

    quickSort(seq,0,N-1)



    prod=1
    a=nCr(N-1,K-1,mod)

    for i in range(1,N-1):
        c=0
        b=0
        if ((N-i-1)>=(K-1)):
            b= nCr(N-i-1,K-1,mod)

        if i>=(K-1) :
            c = nCr(i,K-1,mod)


        a1=power(seq[i],a , mod)
        a2=power(seq[i],(b+c)%1000000006 , mod)
        ans=modDivide(a1,a2,m=1000000007)

        prod=((prod %mod) * (ans % mod))%mod

    print(prod)

File: July2018/NMNMXtemp.cpp
#include <bits/stdc++.h>
#include <algorithm>


using namespace std;

#define ll long long
const int mod = 1e9+7;

ll C[5001][5001];
ll modInv(ll a, ll m=1000000007)
{
    ll m0 = m;
    ll y = 0, x = 1;

    if (m == 1)
      return 0;

    while (a > 1)
    {
        // q is quotient
        ll q = a / m;
        ll t = m;

        // m is remainder now, process same as
        // Euclid's algo
        m = a % m, a = t;
        t = y;

        // Update y and x
        y = x - q * y;
        x = t;
    }

    // Make x positive
    if (x < 0)
       x += m0;

    return x;
}

ll modDivide(ll a, ll b, ll m=1000000007)
{
        a = a % m;
        ll inv = modInv(b,m);
        return (inv*a)%m;
}


int min(int a, int b);

// Returns value of Binomial Coefficient C(n, k)
ll nCr(ll n, ll k)
{
    ll C[n+1][k+1];
    ll i, j;

    // Caculate value of Binomial Coefficient in bottom up manner
    for (i = 0; i <= n; i++)
    {
        for (j = 0; j <= min(i, k); j++)
        {
            // Base Cases
            if (j == 0 || j == i)
                C[i][j] = 1;

            // Calculate value using previosly stored values
            else
                C[i][j] = ((C[i-1][j-1] % 1000000006) + (C[i-1][j] % 1000000006))% 1000000006;
        }
    }

    return (C[n][k] % 1000000006);
}

// A utility function to return minimum of two integers
ll min(ll a, ll b)
{
    return (a<b)? a: b;
}


ll power(ll x, unsigned ll y, ll p)
{
    ll res = 1;      // Initialize result

    x = x % p;  // Update x if it is more than or
                // equal to p

    while (y > 0)
    {
        // If y is odd, multiply x with result
        if (y & 1)
            res = (res*x) % p;

        // y must be even now
        y = y>>1; // y = y/2
        x = (x*x) % p;
    }
    return res;
}


int main()
{

int T = 0;
cin >> T;

while(T)
{
  long int N=0,K=0;
    cin >> N >> K;

    ll seq[N];

for (long int i=0;i<N;i++)
    cin>>seq[i];


    ll n = sizeof(seq)/sizeof(seq[0]);

sort(seq,seq+n);

    ll prod=1;
    ll a=0;
    a=nCr(N-1,K-1);


    for (long int i=1;i<N;i++){
      ll a1=0,a2=0,ans=0;
        ll c=0;
      ll  b=0;
        if ((N-i-1)>=(K-1))
            b= nCr(N-i-1,K-1);

        if (i>=(K-1))
            c = nCr(i,K-1);


        a1=power(seq[i],a , mod);
        a2=power(seq[i],(b+c)%1000000006 , mod);
        ans=modDivide(a1,a2,1000000007);

        prod=((prod %mod) * (ans % mod))%mod;
}
    cout<<prod<<"\n";
    T-=1;
}
return 0;
  }
